### DISCUSSION TITLE
Views and actions on a union

### DISCUSSION THREAD

**Discussion by davetapley** [mobxjs/mobx-state-tree#2081] [Ideas]
It would be cool if I could define `.views` and `.actions` on a `union`.

Currently I have to:

```js
if(Foo.is(foobarpip)) {
   foobarpip.foostuff()
} else if (Bar.is(foobarpip)) {
   foobarpip.barstuff()
} else {
   foobarpip.pipstuff()
}
```

I'd love to just be able to `foo.stuff()` and:

```js
export const FooBarPip = union(Foo, Bar, Pip)
  .actions((self) => ({stuff() { ....
```

---

I _assume_ this would require either a breaking change, or a new type of union altogether, which isn't necessarily a bad thing because there's also https://github.com/mobxjs/mobx-state-tree/issues/1907#issuecomment-1606575347 and:

> Changing the actual behavior would probably require a breaking change

---

Also, is there a more elegant way to `switch` on the concrete type instead of using `if` and  [`is`](https://mobx-state-tree.js.org/API/interfaces/itype#is)?

**Comment by coolsoftwaretyler** [comment]
Hey @davetapley - would the [compose type](https://mobx-state-tree.js.org/API/#compose) fit your use case for defining views and actions on the union of some models? I know it's not exactly what you're looking for, but depending on how you want to use it, maybe that's helpful?

As for a better way to switch, nothing really comes to mind. If you read some of the source where we set up model definitions and handle different primitive and complex types, you'll there's a lot of `if is` going on, albeit with some helpers like `isPrimitive`.

Sorry I can't be more helpful. If you have a proposal for a new type of union, I'd be happy to talk it out. It's not a high priority on the maintainers' end, but there's nothing quite like a well tested and well documented PR to sail through the approval process, especially if you can bolt it on without breaking anything haha.

Sorry for the late response!
